src/Interpreters/InterpreterParallelWithQuery.cpp

```diff
@@ -0,0 +1,199 @@
+#include <Interpreters/InterpreterParallelWithQuery.h>
+
+#include <Core/Settings.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/InterpreterFactory.h>
+#include <Interpreters/executeQuery.h>
+#include <Parsers/queryToString.h>
+#include <Processors/Executors/CompletedPipelineExecutor.h>
+
+
+namespace CurrentMetrics
+{
+    extern const Metric ParallelWithQueryThreads;
+    extern const Metric ParallelWithQueryActiveThreads;
+    extern const Metric ParallelWithQueryScheduledThreads;
+}
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_QUERY;
+}
+
+namespace Setting
+{
+    extern const SettingsMaxThreads max_threads;
+}
+
+
+InterpreterParallelWithQuery::InterpreterParallelWithQuery(const ASTPtr & query_, ContextPtr context_)
+    : WithContext(context_), query(query_), log(getLogger("ParallelWithQuery"))
+{
+}
+
+BlockIO InterpreterParallelWithQuery::execute()
+{
+    const auto & parallel_with_query = query->as<const ASTParallelWithQuery &>();
+    const auto & subqueries = parallel_with_query.children;
+
+    const auto & settings = getContext()->getSettingsRef();
+    size_t max_threads = settings[Setting::max_threads];
+
+    LOG_TRACE(log, "Executing {} subqueries in {} threads", subqueries.size(), max_threads);
+
+    ThreadPoolCallbackRunnerUnsafe<void> schedule;
+    std::unique_ptr<ThreadPool> thread_pool;
+
+    if (max_threads > 1)
+    {
+        thread_pool = std::make_unique<ThreadPool>(
+            CurrentMetrics::ParallelWithQueryThreads,
+            CurrentMetrics::ParallelWithQueryActiveThreads,
+            CurrentMetrics::ParallelWithQueryScheduledThreads,
+            max_threads);
+        schedule = threadPoolCallbackRunnerUnsafe<void>(*thread_pool, "ParallelWithQry");
+    }
```

Comment:
Why do you need to parallel plan creation? It should be fast.

Reply:
There can be statements which do all the work in their interpreter without using any plan at all. For example, `CREATE TABLE`. It's better to execute such things in parallel.